general/mem_manager.hpp

```diff
@@ -67,119 +496,268 @@ class MemoryManager
+}
+
+template <typename T> template <typename U>
+inline Memory<T>::operator const U*() const
+{
+   MFEM_ASSERT(Empty() || (flags & VALID_HOST), "invalid host pointer access");
+   return reinterpret_cast<U*>(h_ptr);
+}
 
-/// Main memory allocation template function. Allocates n*size bytes and returns
-/// a pointer to the allocated memory.
-template<class T>
-inline T *New(const std::size_t n)
+template <typename T>
+inline T *Memory<T>::ReadWriteAccess(MemoryClass mc, int size)
 {
-   T *ptr = new T[n];
-   if (!MemoryManager::Exists()) { return ptr; }
-   return static_cast<T*>(mm.Insert(ptr, n*sizeof(T)));
+   if (!(flags & REGISTERED))
+   {
+      if (mc == MemoryClass::HOST) { return h_ptr; }
+      MemoryManager::Register_(h_ptr, NULL, capacity*sizeof(T),
+                               MemoryType::HOST, flags & OWNS_HOST,
+                               flags & ALIAS, flags);
+   }
+   return (T*)MemoryManager::ReadWriteAccess_(h_ptr, mc, size*sizeof(T), flags);
 }
 
-/// Frees the memory space pointed to by ptr, which must have been returned by a
-/// previous call to mfem::New.
-template<class T>
-inline void Delete(T *ptr)
+template <typename T>
+inline const T *Memory<T>::ReadAccess(MemoryClass mc, int size) const
 {
-   static_assert(!std::is_void<T>::value, "Cannot Delete a void pointer. "
-                 "Explicitly provide the correct type as a template parameter.");
-   if (!ptr) { return; }
-   delete [] ptr;
-   if (!MemoryManager::Exists()) { return; }
-   mm.Erase(ptr);
+   if (!(flags & REGISTERED))
+   {
+      if (mc == MemoryClass::HOST) { return h_ptr; }
+      MemoryManager::Register_((void*)h_ptr, NULL, capacity*sizeof(T),
+                               MemoryType::HOST, flags & OWNS_HOST,
+                               flags & ALIAS, flags);
+   }
+   return (const T *)MemoryManager::ReadAccess_(
+             (void*)h_ptr, mc, size*sizeof(T), flags);
 }
 
-/// Return a host or device address corresponding to current memory space
-template <class T>
-inline T *Ptr(T *a) { return static_cast<T*>(mm.Ptr(a)); }
+template <typename T>
+inline T *Memory<T>::WriteAccess(MemoryClass mc, int size)
+{
+   if (!(flags & REGISTERED))
+   {
+      if (mc == MemoryClass::HOST) { return h_ptr; }
+      MemoryManager::Register_(h_ptr, NULL, capacity*sizeof(T),
+                               MemoryType::HOST, flags & OWNS_HOST,
+                               flags & ALIAS, flags);
+   }
+   return (T*)MemoryManager::WriteAccess_(h_ptr, mc, size*sizeof(T), flags);
+}
 
-/// Data will be pushed/pulled before the copy happens on the host or the device
-inline void* Memcpy(void *dst, const void *src,
-                    std::size_t bytes, const bool async = false)
-{ return mm.Memcpy(dst, src, bytes, async); }
+template <typename T>
+inline void Memory<T>::SyncWith(const Memory &other) const
+{
+   if (!(flags & REGISTERED) && (other.flags & REGISTERED))
+   {
+      MFEM_ASSERT(h_ptr == other.h_ptr &&
```

Comment:
Why is this an invalid input?

Reply:
The reasoning for the assert comes from:
1) This method is meant to work in two cases: either `h_ptr == other.h_ptr` or `*this` is an alias of `other`.
2) Alias `Memory` is not allowed to be registered with the `MemoryManager` after construction.

There could be another assert outside the if-statement but for now I just wanted to make sure that if `*this` is marked as registered, it is done in cases where it makes sense.